Sources/Sentry/SentryTracer.m

```diff
@@ -103,13 +64,43 @@ - (instancetype)initWithParentTracer:(SentryTracer *)parent context:(SentrySpanC
                                            sampled:_rootSpan.context.sampled];
     context.spanDescription = description;
 
-    SentryTracer *child = [[SentryTracer alloc] initWithParentTracer:self context:context];
-    @synchronized([self childrenLock]) {
-        [_spans addObject:child];
+    SentrySpan *child = [[SentrySpan alloc] initWithTracer:self context:context];
+
+    if (_waitForChildren) {
+        // Observe when the child finishes
+        [child addObserver:self
+                forKeyPath:NSStringFromSelector(@selector(timestamp))
+                   options:NSKeyValueObservingOptionNew
+                   context:&spanTimestampObserver];
     }
+
+    @synchronized(_children) {
+        [_children addObject:child];
+    }
+
     return child;
 }
 
+/**
+ * Is called when a span finishes and checks if we can finish.
+ */
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey, id> *)change
+                       context:(void *)context
+{
+    if (context == spanTimestampObserver &&
+        [keyPath isEqualToString:NSStringFromSelector(@selector(timestamp))]) {
```

Comment:
Checking context and keyPath here is redundant.

Reply:
Thanks for the info.